pkg/deploy/depoy.go

```diff
@@ -0,0 +1,247 @@
+	location          string
+}
+
+// New initiates new deploy utility object
+func New(ctx context.Context, log *logrus.Entry, authorizer autorest.Authorizer, gitCommit string) (*deployer, error) {
+	d := &deployer{
+		log:               log,
+		version:           gitCommit,
+		resourceGroupName: os.Getenv("AZURE_RP_RESOURCEGROUP_NAME"),
+		subscriptionID:    os.Getenv("AZURE_SUBSCRIPTION_ID"),
+	}
+
+	d.cli = &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	// extract location from rg
+	gCli := resources.NewGroupsClient(d.subscriptionID, authorizer)
+	group, err := gCli.Get(ctx, d.resourceGroupName)
+	if err != nil {
+		return nil, err
+	}
+	d.location = *group.Location
+
+	d.deployments = resources.NewDeploymentsClient(d.subscriptionID, authorizer)
+	d.vmss = compute.NewVirtualMachineScaleSetsClient(d.subscriptionID, authorizer)
+	d.vmssvms = compute.NewVirtualMachineScaleSetVMsClient(d.subscriptionID, authorizer)
+	d.network = network.NewPublicIPAddressesClient(d.subscriptionID, authorizer)
+
+	return d, nil
+}
+
+// PreDeploy deploys NSG and ManagedIdentity, needed for man deployment
+func (d *deployer) PreDeploy(ctx context.Context, log *logrus.Entry) (rpServicePrincipalID string, err error) {
+	// if RP_PREDEPLOY is not set, we read old rpServicePrincipalID and return it
+	if len(os.Getenv("AZURE_RP_PREDEPLOY")) != 0 {
+		var data []byte
+		data, err = DeployRpProductionNsgJsonBytes()
+		if err != nil {
+			return
+		}
+
+		var azuretemplate map[string]interface{}
+		err = json.Unmarshal(data, &azuretemplate)
+		if err != nil {
+			return
+		}
+
+		log.Infof("deploying nsg and managedIdentity to %s", d.resourceGroupName)
+		err = d.deployments.CreateOrUpdateAndWait(ctx, d.resourceGroupName, "rp-production-nsg", azresources.Deployment{
+			Properties: &azresources.DeploymentProperties{
+				Template: azuretemplate,
+				Mode:     azresources.Incremental,
+			},
+		})
+		if err != nil {
+			return
+		}
+	}
+	deployment, err := d.deployments.Get(ctx, d.resourceGroupName, "rp-production-nsg")
+	if err != nil {
+		return
+	}
+	return deployment.Properties.Outputs.(map[string]interface{})["rpServicePrincipalId"].(map[string]interface{})["value"].(string), nil
+}
+
+func (d *deployer) Deploy(ctx context.Context, log *logrus.Entry, rpServicePrincipalID string) error {
+	data, err := DeployRpProductionJsonBytes()
+	if err != nil {
+		return err
+	}
+
+	var azuretemplate map[string]interface{}
+	err = json.Unmarshal(data, &azuretemplate)
+	if err != nil {
+		return err
+	}
+
+	parameters, err := getParameters()
+	if err != nil {
+		return err
+	}
+	parameters["vmssName"].(map[string]interface{})["value"] = d.version
+	parameters["rpServicePrincipalId"].(map[string]interface{})["value"] = rpServicePrincipalID
+	// azure enforce ^[a-z][a-z0-9-]{1,61}[a-z0-9]$. for DomainNameLabel.
+	// gitCommit do not comply as it might start with a number
+	parameters["vmssDomainNameLabel"].(map[string]interface{})["value"] = "vmss-" + d.version
```

Comment:
I think you can make this less unpleasant by using the `arm.Parameters` type throughout.  Perhaps at the end you need to manually marshal it into a []byte and give that to azresources.DeploymentProperties; perhaps not.